lib/rspec-puppet/adapters.rb

```diff
@@ -237,8 +237,8 @@ def setup_puppet(example_group)
         case RSpec.configuration.facter_implementation.to_sym
         when :rspec
           if supports_facter_runtime?
-            Puppet.runtime[:facter] = proc { RSpec::Puppet::FacterTestImpl.new }
-            set_facter_impl(Puppet.runtime[:facter])
+            set_facter_impl(RSpec::Puppet::FacterTestImpl.new)
```

Comment:
I know I suggested this, but after posting I'm not sure if it needs to be a `proc` or not. Now it's a singleton where previously it wasn't.

Some testcase that comes to mind:
```ruby
describe 'myclass' do
  let(:facts) do
    {myfact: 'value'}
  end

  it { ... } # verify only myfact is set

  context 'with additional fact' do
    let(:facts) { super().merge(otherfact: 'some value') }

    it { ... } # verify both are set is set
  end

  context 'fresh facts' do
    let(:facts) do
      {otherfact: 'some value'}
    end

    it { ... } # verify only otherfact is set is set
  end
end
```

Reply:
Indeed `RSpec::Puppet::FacterTestImpl.new` was lazily allocated, but the proc was evaluated on line 241 in the call to `Puppet.runtime[:facter]` ([implementation](https://github.com/puppetlabs/puppet/blob/7dbd0d997d4e6ef600e9c4d535efc647c739ebae/lib/puppet/runtime.rb#L44)); so it didn't make much sense to use a proc.

Technically we can shave off some of the allocated memory if we use a proc and do something like this:

```ruby
def set_facter_impl(impl)
  unless defined? FacterImpl
    impl = impl.call if impl.is_a?(Proc)
    Object.send(:const_set, :FacterImpl, impl)
  end
end

set_facter_impl(proc { RSpec::Puppet::FacterTestImpl.new })
Puppet.runtime[:facter] = FacterImpl
```

This way the `FacterTestImpl` object will only be instantiated if the `FacterImpl` constant is not defined, and should make things a bit faster.

Reply:
You're right. That was not immediately obvious to me (it may be because I was on PTO for 2 weeks) but the difference is that `Facter` itself already the "instance" (since it implements the methods as static) to call on where `FacterTestImpl` is a class.

The alternative is to enhance `FacterTestImpl` to implements the methods statically by (lazily) instantiating itself but I'm not sure if that's really better than this.

I'm fine with your suggestion but I'd appreciate a small description of that difference. Either in the commit message or as a code comment.

Reply:
> That was not immediately obvious to me (it may be because I was on PTO for 2 weeks)

I just came back from a 2 week PTO myself, so I totally get that 😆